internal/smb2/dtyp.go

```diff
@@ -148,3 +214,264 @@ func (c SidDecoder) Decode() *Sid {
 		SubAuthority:        c.SubAuthority(),
 	}
 }
+
+type SecurityDescriptorDecoder []byte
+
+func (c SecurityDescriptorDecoder) IsInvalid() bool {
+	if len(c) < 20 {
+		return true
+	}
+
+	expLength := 20
+	if c.Revision() != 1 {
+		return true
+	}
+	if c.OffsetOwner() != 0 {
+		if c.OwnerSid().IsInvalid() {
+			return true
+		}
+		expLength += c.OwnerSid().Decode().Size()
+	}
+
+	if c.OffsetGroup() != 0 {
+		if c.GroupSid().IsInvalid() {
+			return true
+		}
+		expLength += c.GroupSid().Decode().Size()
+	}
+
+	if c.OffsetSacl() != 0 {
+		if c.Control()&SECURITY_DESCRIPTOR_SACL_PRESENT == 0 || c.Sacl().IsInvalid() {
+			return true
+		}
+		expLength += int(c.Sacl().AclSize())
+	} else if c.Control()&SECURITY_DESCRIPTOR_SACL_PRESENT != 0 {
+		return true
+	}
+
+	if c.OffsetDacl() != 0 {
+		if c.Control()&SECURITY_DESCRIPTOR_DACL_PRESENT == 0 || c.Dacl().IsInvalid() {
+			return true
+		}
+		expLength += int(c.Dacl().AclSize())
+	} else if c.Control()&SECURITY_DESCRIPTOR_DACL_PRESENT != 0 {
+		return true
+	}
+
+	if len(c) != expLength {
+		return true
+	}
+
+	return false
+}
+
+func (c SecurityDescriptorDecoder) Revision() uint8 {
+	return c[0]
+}
+
+func (c SecurityDescriptorDecoder) Sbz1() uint8 {
+	return c[1]
+}
+
+func (c SecurityDescriptorDecoder) Control() uint16 {
+	return le.Uint16(c[2:4])
+}
+
+func (c SecurityDescriptorDecoder) OffsetOwner() uint32 {
+	return le.Uint32(c[4:8])
+}
+
+func (c SecurityDescriptorDecoder) OffsetGroup() uint32 {
+	return le.Uint32(c[8:12])
+}
+
+func (c SecurityDescriptorDecoder) OffsetSacl() uint32 {
+	return le.Uint32(c[12:16])
+}
+
+func (c SecurityDescriptorDecoder) OffsetDacl() uint32 {
+	return le.Uint32(c[16:20])
+}
+
+func (c SecurityDescriptorDecoder) OwnerSid() SidDecoder {
+	return sidDecoderFromBytes(c[c.OffsetOwner():])
+}
+
+func (c SecurityDescriptorDecoder) GroupSid() SidDecoder {
+	return sidDecoderFromBytes(c[c.OffsetGroup():])
+}
+
+func (c SecurityDescriptorDecoder) Sacl() ACLDecoder {
+	if c.Control()&SECURITY_DESCRIPTOR_SACL_PRESENT == 0 {
+		return nil
+	}
+	return aclDecoderFromBytes(c[c.OffsetSacl():])
+}
+
+func (c SecurityDescriptorDecoder) Dacl() ACLDecoder {
+	if c.Control()&SECURITY_DESCRIPTOR_DACL_PRESENT == 0 {
+		return nil
+	}
+	return aclDecoderFromBytes(c[c.OffsetDacl():])
+}
+
+func sidDecoderFromBytes(b []byte) SidDecoder {
```

Comment:
I guess this wrapper isn't necessary if we always check IsInvalid before using SidDecoder.

Reply:
I'm using it as a helper function since I don't know the size of the Sid. I'm not really sure how IsInvalid could change this.
I could inline the helper function! :slightly_smiling_face:

Reply:
what happen if we replace all `sidDecoderFromBytes(b)` into `SidDecoder(b)`?
If we don't see anything broken, that's unnecessary.